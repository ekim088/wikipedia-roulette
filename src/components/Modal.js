// @flow
import * as React from 'react';
import { useEffect, useRef } from 'react';
import { CSSTransition } from 'react-transition-group';
import '../scss/_Modal.scss';

type Props = {
	children?: React.Node | string,
	onHide: ({ ... }) => void,
	show: boolean
};

const Modal = ({ children, onHide, show }: Props) => {
	const keyboadEvent = 'keyup';
	const mouseEvent = 'click';
	const modalRef: { current: any } = useRef(null);

	useEffect(() => {
		/**
		 * Call `onHide` handler from parent composite component on press of
		 * `escape` key.
		 * @param {Event} e The click event.
		 */
		const handleKeyUp = (e: KeyboardEvent) => {
			// toggle on escape
			if (e.keyCode === 27) {
				e.preventDefault();
				onHide(e);
			}
		};

		/**
		 * Call `onHide` handler from parent composite component when clicking
		 * on background overlay surrounding dialog.
		 * @param {Event} e The click event.
		 */
		const handleBackgroundClick = (e: MouseEvent) => {
			if (modalRef.current && !modalRef.current.contains(e.target)) {
				onHide(e);
			}
		};

		// apply/remove post page load event handlers to toggle modal
		if (show) {
			if (modalRef.current) modalRef.current.focus();
			window.addEventListener(keyboadEvent, handleKeyUp);
			document.addEventListener(mouseEvent, handleBackgroundClick);
		} else {
			window.removeEventListener(keyboadEvent, handleKeyUp);
			document.removeEventListener(mouseEvent, handleBackgroundClick);
		}

		return () => {
			// unbind event handlers on unnmount
			window.removeEventListener(keyboadEvent, handleKeyUp);
			document.removeEventListener(mouseEvent, handleBackgroundClick);
		};
	}, [onHide, show]);

	return (
		<CSSTransition in={show} classNames="wa-modal" timeout={300} unmountOnExit>
			<div className="wa-modal-container" key="wa-modal">
				<dialog
					className="wa-modal"
					ref={modalRef}
					{...(show ? { open: true } : {})}
				>
					<button type="button" className="wa-modal-close" onClick={onHide}>
						X
					</button>
					<div>{children}</div>
				</dialog>
			</div>
		</CSSTransition>
	);
};

Modal.defaultProps = {
	children: undefined
};

export default Modal;
