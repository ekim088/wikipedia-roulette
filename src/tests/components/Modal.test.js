import Modal from '../../components/Modal';

describe('Modal', () => {
	const documentAddEventListenerFxn = document.addEventListener;
	const documentRemoveEventListenerFxn = document.removeEventListener;
	const windowAddEventListenerFxn = window.addEventListener;
	const windowRemoveEventListenerFxn = window.removeEventListener;
	let eventMap = {};
	const eventMapBasedListener = (event, callback) => {
		eventMap[event] = callback;
	};

	afterEach(() => {
		// reset event map for simulating bindings and events
		eventMap = {};

		// reset mocks to document and window methods
		document.addEventListener = documentAddEventListenerFxn;
		document.removeEventListener = documentRemoveEventListenerFxn;
		window.addEventListener = windowAddEventListenerFxn;
		window.removeEventListener = windowRemoveEventListenerFxn;
	});

	it('should render a modal with a close button and content div', () => {
		const wrapper = shallow(<Modal show onHide={jest.fn()} />);
		expect(wrapper).toMatchSnapshot();
	});

	it('should call the onHide method on click of the close button', () => {
		const mockOnHide = jest.fn();
		shallow(<Modal show onHide={mockOnHide} />)
			.find('button')
			.simulate('click');
		expect(mockOnHide).toHaveBeenCalled();
	});

	it('should setup event listeners to manage modal on mount', () => {
		window.addEventListener = jest.fn();
		document.addEventListener = jest.fn();
		mount(<Modal show onHide={jest.fn()} />);
		expect(window.addEventListener).toHaveBeenCalledWith(
			'keyup',
			expect.any(Function)
		);
		expect(document.addEventListener).toHaveBeenCalledWith(
			'click',
			expect.any(Function)
		);
	});

	it('should remove event listeners to manage modal on unmount', () => {
		window.removeEventListener = jest.fn();
		document.removeEventListener = jest.fn();
		mount(<Modal show onHide={jest.fn()} />).unmount();
		expect(window.removeEventListener).toHaveBeenCalledWith(
			'keyup',
			expect.any(Function)
		);
		expect(document.removeEventListener).toHaveBeenCalledWith(
			'click',
			expect.any(Function)
		);
	});

	it('should trigger onHide function when clicking outside of the modal', () => {
		/**
		 * Setup event map to simulate document events (enzyme simulate() will
		 * not work with document events)
		 */
		document.addEventListener = jest.fn(eventMapBasedListener);
		document.removeEventListener = jest.fn();
		const onHide = jest.fn();
		const wrapper = mount(<Modal show onHide={onHide} />);

		// simulate click inside modal
		const modalNode = wrapper.find('.wa-modal').getDOMNode();
		eventMap.click({ target: modalNode });
		expect(onHide).not.toHaveBeenCalled();

		// simulate click outside modal
		const modalContainerNode = wrapper.find('.wa-modal-container').getDOMNode();
		const mockEventWithOutsideEl = { target: modalContainerNode };
		eventMap.click(mockEventWithOutsideEl);
		expect(onHide).toHaveBeenCalledWith(mockEventWithOutsideEl);
	});

	it('should trigger onHide function when pressing escape key', () => {
		/**
		 * Setup event map to simulate document events (enzyme simulate() will
		 * not work with document events)
		 */
		window.addEventListener = jest.fn(eventMapBasedListener);
		window.removeEventListener = jest.fn();
		const onHide = jest.fn();
		mount(<Modal show onHide={onHide} />);

		// simulate enter key press
		eventMap.keyup({ keyCode: 13 });
		expect(onHide).not.toHaveBeenCalled();

		// simulate escape key press
		const mockEvent = {
			keyCode: 27,
			preventDefault: jest.fn()
		};
		eventMap.keyup(mockEvent);
		expect(onHide).toHaveBeenCalledWith(mockEvent);
		expect(mockEvent.preventDefault).toHaveBeenCalled();
	});
});
