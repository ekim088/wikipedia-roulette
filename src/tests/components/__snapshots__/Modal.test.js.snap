// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render a modal with a close button and content div 1`] = `
<CSSTransition
  classNames="wa-modal"
  in={true}
  timeout={300}
  unmountOnExit={true}
>
  <div
    className="wa-modal-container"
    key="wa-modal"
  >
    <dialog
      className="wa-modal"
      open={true}
    >
      <button
        className="wa-modal-close"
        onClick={[MockFunction]}
        type="button"
      >
        X
      </button>
      <div />
    </dialog>
  </div>
</CSSTransition>
`;
